crates/types/src/partitions/state.rs

```diff
@@ -0,0 +1,182 @@
+    ) -> bool {
+        let mut modified = false;
+        match incoming_current_membership
+            .version
+            .cmp(&self.observed_current_membership.version)
+        {
+            // we have a new current membership
+            std::cmp::Ordering::Greater => {
+                self.observed_current_membership = incoming_current_membership.clone();
+                modified = true;
+                if self
+                    .observed_next_membership
+                    .as_ref()
+                    .is_some_and(|my_next| {
+                        my_next.version <= self.observed_current_membership.version
+                    })
+                {
+                    // unset our next, our current has caught up
+                    self.observed_next_membership = None;
+                }
+            }
+            std::cmp::Ordering::Equal => {
+                // merge member's durable lsns
+                modified = self
+                    .observed_current_membership
+                    .merge(incoming_current_membership.clone());
+            }
+            std::cmp::Ordering::Less => { /* ignore it */ }
+        }
+
+        // dealing with next membership configuration
+        let Some(incoming_next_membership) = incoming_next_membership else {
+            return modified;
+        };
+
+        // incoming has next but it older/equal to our own current
+        if incoming_next_membership.version <= self.observed_current_membership.version {
+            // ignore it, their next is lower that our current's
+            return modified;
+        }
+
+        let Some(my_next_membership) = &mut self.observed_next_membership else {
+            self.observed_next_membership = Some(incoming_next_membership.clone());
+            return true;
+        };
+
+        match incoming_next_membership
+            .version
+            .cmp(&my_next_membership.version)
+        {
+            std::cmp::Ordering::Greater => {
+                *my_next_membership = incoming_next_membership.clone();
+                modified = true;
+            }
+            std::cmp::Ordering::Equal => {
+                modified = my_next_membership.merge(incoming_next_membership.clone());
+            }
+            std::cmp::Ordering::Less => { /* ignore it */ }
+        }
+
+        modified
+    }
+}
+
+#[derive(Debug, Clone, bilrost::Message, NetSerde)]
+pub struct ReplicaSetState {
+    pub version: Version,
+    // ordered, akin to NodeSet
+    pub members: Vec<MemberState>,
+}
+
+impl Default for ReplicaSetState {
+    fn default() -> Self {
+        Self {
+            version: Version::INVALID,
+            members: Vec::new(),
+        }
+    }
+}
+
+impl Merge for ReplicaSetState {
+    fn merge(&mut self, other: Self) -> bool {
+        debug_assert_eq!(
+            self.members.iter().map(|m| m.node_id).collect::<Vec<_>>(),
+            other.members.iter().map(|m| m.node_id).collect::<Vec<_>>(),
```

Comment:
nit: one could use `itertools::assert_equal` to avoid the vector allocation

Reply:
That's a nice one, it's good one to know for cases that are not debug_assertion gated. For this one, I'd opt to not include itertools dependency in this module unless I have other uses for it.